Migrate the timeline data layer to TypeScript

reykjalin opened the pull request:
Fixes #1275 

Depends on #1279 and #1263.

Adds types to most things in `client/data/timeline`. Unfortunately we won't see everything come together neatly until the whole data store has been migrated, so some of the type information is still a bit limited.

#### Changes proposed in this Pull Request

- Migrate the timeline data store to TypeScript.

#### Testing instructions

- `rm -rf dist`
- `npm ci`
- `npm run test:js`
- Make sure tests run successfully
- `npm run watch`
- Make sure payment details screens load properly.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

Commit:
Add typechecking to JS lint script

Commit:
WIP: migrate charges data layer

Commit:
WIP: migrate charges data layer

Commit:
Add types to charge hooks

Commit:
Migrate rest of data store to TypeScript

Commit:
Fill out most of the charge data type

Commit:
Split charge state into errors and data

Commit:
Remove unnecessary data from charge actions

Commit:
Migrate timeline reducer to TypeScript

Commit:
Migrate timeline hooks to TS

Commit:
Migrate rest of timeline data layer to TS

Commit:
Migrate tests to TS

Commit:
clear error state when charge is updated

Commit:
Inline type check npm script

Commit:
Flesh out Timeline types

Commit:
add missing types for @wordpress/data-controls

Commit:
Improve payment method details typing

Commit:
Make billing details required on Charge objects

vbelolapotkov commented on client/data/charges/types.ts:
```diff
@@ -0,0 +1,172 @@
+	wechat?: unknown;
+}
+
+export interface Charge {
```
@reykjalin what was your approach for building this interface? I've tried to compare it with [Stripe charge object](https://stripe.com/docs/api/charges/object) and can see some connect specific fields are missing there.

Have you also seen [stripe-node lib](https://github.com/stripe/stripe-node)? We might use types from it or copy from it. What do you think?

reykjalin replied:
I referenced the Stripe API and our own code. I did know of the Stripe lib, but decided to define our own type here instead because we may modify the response before passing it on to the frontend.

There's a discussion on exactly this in https://github.com/Automattic/woocommerce-payments/pull/1279#discussion_r577477482 and I think it'd be best to keep this discussion there.

I'd love to hear more of your thoughts on this there!

vbelolapotkov commented on client/data/charges/types.ts:
```diff
@@ -0,0 +1,172 @@
+/* eslint-disable camelcase */
+export interface Address {
```
How about setting all fields as required but use [Parital utility](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype)?

reykjalin replied:
I think the `Partial<T>` type is really only useful for function declarations. The issue here is that Stripe (or even our API) may return `null` for each of these fields which is why they're optional.

The type interface here doesn't have anything to do with using the type, it's more about defining the structure of the type. The `Partial<T>` utility is intended for when you need to use a variable that might only contain partial information for a type `T`.

vbelolapotkov replied:
Fair point. Depending on the outcome of conversation in https://github.com/Automattic/woocommerce-payments/pull/1279#discussion_r577477482 we might still end up with something like `interface WCPayType = Partial<StripeType>`.

vbelolapotkov commented on client/data/timeline/hooks.ts:
```diff
@@ -0,0 +1,66 @@
+			} = select( STORE_NAME );
+
+			// Make sure the charge is loaded first to get the intention ID.
+			const isChargeLoading = isResolving< boolean >( 'getCharge', [
```
That's probably a styling preference, but IMO this way it reads a bit better. What do you think?
```suggestion
			const isChargeLoading: boolean = isResolving( 'getCharge', [
```

reykjalin replied:
The issue here is more in that the `select()` method returns [generic functions](https://www.typescriptlang.org/docs/handbook/generics.html#hello-world-of-generics) so you must type the function in the angle brackets.

I agree your suggestion reads much better, and your suggestion works for that specific case, but unfortunately doing it that way results in an error on a different line:

<img width="741" alt="image" src="https://user-images.githubusercontent.com/13835680/109343048-b2d35d00-78a7-11eb-8fb9-2c3b56df4103.png">

whereas typing the function in the angle bracket provides the right result:

<img width="956" alt="image" src="https://user-images.githubusercontent.com/13835680/109343166-dac2c080-78a7-11eb-9bb2-ca55b21f6768.png">

I used angle brackets everywhere for the sake of consistency in the code here 🙂

vbelolapotkov replied:
Good point, agree!

vbelolapotkov commented on client/data/timeline/hooks.ts:
```diff
@@ -0,0 +1,66 @@
+			const isChargeLoading = isResolving< boolean >( 'getCharge', [
+				chargeId,
+			] );
+			const chargeError = getChargeError< Error | undefined >( chargeId );
```
Looking at those `Type | undefined` I want to add something like `Maybe<T> = T | undefined`  and reuse it

reykjalin replied:
I understand that reaction, but I think in this case the `Type | undefined` is clearer. I think using a `Maybe<T>` type will result in people thinking there's more behind it, e.g. a maybe monad (common in functional languages) which comes with its own interface.

I think using a `Maybe<T>` type will just end up causing confusion, so I'm hesitant to recommending that 😬

vbelolapotkov replied:
Is there a chance to use better naming than `Mabye<T>`, e.g. `Optional<T>`, `UndefinedOr<T>`, etc. it feels weird to me carry all those `undefined`s.

reykjalin replied:
I honestly don't see much value in creating a wrapper type like that. It's just a different way to do the same thing and doesn't benefit us much IMO.

Besides, I'm pretty sure the TS community generally uses variations of `<type> | undefined` (e.g. `string | undefined`) instead of creating wrapper types.

reykjalin replied:
For this specific case `getChargeError< Error | undefined >( /* ... */ );` isn't all that different from `getChargeError< UndefinedOr< Error > >( /* ... */ )` and is really just slightly different noise 🙂

vbelolapotkov commented on client/data/timeline/resolvers.ts:
```diff
@@ -10,13 +10,25 @@ import { apiFetch } from '@wordpress/data-controls';
 
-export function* getTimeline( intentionId ) {
+const isTimeline = ( responseData: any ): responseData is Timeline => {
+	return (
+		typeof responseData === 'object' &&
+		undefined !== ( responseData as Timeline ).length
+	);
```
Same as with charges, I don't fully understand the purpose of this check

reykjalin replied:
It would be best to refer to my other reply: https://github.com/Automattic/woocommerce-payments/pull/1280#discussion_r583842155

To be fair though; I'm not sure this is the best way to do this type guard. I'd love to get some recommendations there.

vbelolapotkov replied:
Well, the decision is the same as above. So it's fine to go as is, I'm only curious what happens to UX if this check fails. I.e. neither update nor error action are triggered. Will it stuck in loading state? If so we'd probably need to generate/throw error in else.

reykjalin replied:
As per previous discussions, [we now throw an error](https://github.com/Automattic/woocommerce-payments/blob/6d0a079b71a903357d47f69c4238258dd4bc8281/client/data/charges/resolvers.ts#L27-L29) when the type guard fails 🙂

vbelolapotkov commented on client/data/timeline/selectors.ts:
```diff
@@ -0,0 +1,21 @@
+export const getTimelineError = (
+	state: { timeline: TimelineState },
+	id: string
+): Error | undefined => {
```
This changes the signature of the result a bit. Previously we'd return empty object if error is missing. I haven't tested site errors, so not 100% sure if it affects the UI.

reykjalin replied:
I chose to go with `undefined` instead of empty object because the type signature of an empty object is ugly; `Record<string, never>`.

I think it will be better for us to use `undefined` here as well because it's clearer what's going on; there's no error, not an empty error.

I do agree that we have to be careful here though, but I'm not sure how to test the error handling in the UI. Do I have to just change the code to return malformed data form the server?

vbelolapotkov replied:
I like the change but it should be covered with corresponding test IMO.

> I do agree that we have to be careful here though, at least until everything has been migrated, but I'm not sure how to test the error handling in the UI. Do I have to just change the code to return malformed data form the server?

Yep, either throwing an error on server. Or if working in test set mock to throw error.

reykjalin replied:
> it should be covered with a corresponding test

[It already is](https://github.com/Automattic/woocommerce-payments/blob/6d0a079b71a903357d47f69c4238258dd4bc8281/client/data/timeline/test/selectors.ts#L50-L60) 🙂

vbelolapotkov commented on client/data/timeline/types.ts:
```diff
@@ -0,0 +1,145 @@
+/* eslint-disable camelcase */
```
Haven't checked each field, but thoughts are the same as with types for charge, maybe worth checking stripe lib for this.

reykjalin replied:
The Timeline is entirely our own thing, so I think we have to define our own types here.

That said, your point still stands; we might be able to use some of Stripe's pre-defined types here, but I'm personally leaning to managing our own types.

vbelolapotkov replied:
Agree here unless we proxy some stripe objects (haven't checked this). Did you check?

reykjalin replied:
Yep, I did check. We do proxy some objects, but mostly we restructure the data to better fit the timeline, sometimes extracting just specific parts of an object.

Mostly the Timeline data structure is our own.

vbelolapotkov reviewed (commented):
Hey @reykjalin ! Great progress on TS! I left some comments below.

Commit:
Use null-coalescing operator instead of `||`

Commit:
handle invalid charge response from store API

Commit:
remove redundant undefined check

reykjalin commented on client/data/charges/reducer.ts:
```diff
@@ -0,0 +1,57 @@
+	return state;
+};
+
+const reducer = combineReducers( {
+	data,
+	errors,
+} );
```
Yes, it is!

The implementation was made a bit clearer in terms of the typing by changing the shape and splitting up the data store. I drew heavy inspiration from the calypso data stores IIRC 🙂

vbelolapotkov replied:
Well done! I like it. Worth a p2 post IMO ;)

reykjalin commented on client/data/charges/resolvers.ts:
```diff
@@ -0,0 +1,34 @@
+			path: `${ NAMESPACE }/charges/${ id }`,
+		} );
+
+		if ( isCharge( results ) ) {
```
This is a fantastic question!

This is what's called a "type guard" and you can read more about them [in the TypeScript handbook](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards).

This is necessary to confirm that what's returned from `apiFetch()` is, in fact, a `Charge` object. Once `isCharge()` returns `true`, the type system will know for sure that `results` is a `Charge` object.

If the API returns something that's not a `Charge` (and we return `false` from `isCharge`) the type system will know that it's not a `Charge` and we need to explicitly handle that. I've done that in the `migrate/data-charges` branch, but haven't rebased this one yet 🙂

vbelolapotkov replied:
Thanks for explaining this Kris! Yeah, I've heard of the type guards and old me would do something simple like [type assertion](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions): `const results = yield apiFetch(...) as Charge;` but this way makes sense too.

reykjalin commented on client/data/charges/resolvers.ts:
```diff
@@ -0,0 +1,34 @@
+
+const isCharge = ( apiResponse: any ): apiResponse is Charge => {
+	return (
+		( apiResponse as Charge ).object !== undefined &&
```
You're right! The `'charge'` equivalence check is enough here. This has been fixed in #1279 ☺️

vbelolapotkov replied:
👍

reykjalin commented on client/data/charges/types.ts:
```diff
@@ -0,0 +1,172 @@
+
+export interface CardPaymentMethodDetails {
+	brand:
+		| 'amex'
+		| 'diners'
+		| 'discover'
+		| 'jcb'
+		| 'mastercard'
+		| 'unionpay'
+		| 'visa'
+		| 'unknown';
```
Hmmm, the end result will probably be similar, but marking this as a type is more semantically correct.

I've changed this a bit in #1279 and I think that does a better job of explaining the benefits here. I think in essence it's probably easiest to say that using a [literal type](https://www.typescriptlang.org/docs/handbook/literal-types.html) will help the type system more than a string enum.

I understand that's not a great answer to the question, so please let me know if there's anything I can clarify!

vbelolapotkov replied:
TBH, I don't have preference here and after a quick search it's still unclear when literal vs enum is preferred. So I trust your choice here, but in the end it'd be nice to see a post with a list of such decisions/recommendations so we stay consistent when others start working in TS :)

reykjalin replied:
I think the general approach should be "use a `type` whenever you can" :)

Not sure if that's something we should document. We're using TypeScript to get access to a proper type system; we should use it :)

reykjalin commented on client/data/charges/types.ts:
```diff
@@ -0,0 +1,172 @@
+	amount: number;
+	amount_captured: number;
+	amount_refunded: number;
+	balance_transaction: string;
```
Yes, you're right! I didn't think we ever expanded the object which is why I left is as a `string` here. This should indeed be `string | BalanceTransaction`, and I'll need to write out the `BalanceTransaction` type 🙂 

Great catch! 👏

reykjalin replied:
Fixed in 7f426f97e8bc9df8d849db0e354d7d3efa39a536 🙂

reykjalin commented on client/data/charges/types.ts:
```diff
@@ -0,0 +1,172 @@
+	status: 'pending' | 'succeeded' | 'failed';
+}
+
+export interface ChargesStateEntry {
+	data?: Charge;
+	error?: Error;
+}
```
Ohh, thanks for catching that! I forgot to remove this after changing the shape of the data store ❤️

reykjalin replied:
Fixed in c985252dea9edb6b4e60be69a16fe2b8a5bfe94c ❤️

Commit:
pin @types/wordpress__data-controls package version

Commit:
extend Stripe JS types instead of completely rolling our own

Commit:
add balance transaction type

Commit:
remove unused type

Commit:
Improve the Timeline type guard

This is a very minor improvement, mostly in terms of readability since
we know just check whether the data from the API is an array.

Commit:
add stripe API specific type declarations

Commit:
use Stripe API specific types

Commit:
add possible nulls to types

Commit:
fix outdated comment

reykjalin commented:
Too old to keep working on, so closing this.
